File: flutter_app/ios/Runner/EzvizSdk.swift
import Foundation

final class EzvizSdk {
  static let shared = EzvizSdk()
  private init() {}

  private var initialized = false
  private var initializedWithCustomServer = false
  private var lastApiUrl: String?
  private var lastAuthUrl: String?

  private func normalizeBaseUrl(_ raw: String) -> String {
    guard let u = URL(string: raw), let scheme = u.scheme, let host = u.host else {
      return raw
    }
    // EZOpenSDK expects base host, it will append API paths internally.
    return "\(scheme)://\(host)"
  }

  func initIfNeeded(appKey: String, accessToken: String?, apiUrl: String?, authUrl: String?) {
    guard !appKey.isEmpty else { return }
    let wantCustom = (apiUrl?.isEmpty == false) && (authUrl?.isEmpty == false)
    let normApiUrl = apiUrl.map(normalizeBaseUrl)
    let normAuthUrl = authUrl.map(normalizeBaseUrl)

    // If we already initialized with default server but later user provides custom server,
    // we must re-init, otherwise the SDK keeps using the default CN endpoints.
    if initialized && wantCustom && !initializedWithCustomServer {
      _ = EZOpenSDK.destoryLib()
      initialized = false
    }

    if !initialized {
      // Must be called before init.
      EZOpenSDK.setDebugLogEnable(false)
      if wantCustom, let apiUrl = normApiUrl, let authUrl = normAuthUrl {
        lastApiUrl = apiUrl
        lastAuthUrl = authUrl
        initialized = EZOpenSDK.initLib(withAppKey: appKey, url: apiUrl, authUrl: authUrl)
        initializedWithCustomServer = initialized
      } else {
        lastApiUrl = nil
        lastAuthUrl = nil
        initialized = EZOpenSDK.initLib(withAppKey: appKey)
        initializedWithCustomServer = false
      }
    } else if wantCustom {
      // Keep for debugging.
      lastApiUrl = normApiUrl
      lastAuthUrl = normAuthUrl
    }
    if let token = accessToken, !token.isEmpty {
      EZOpenSDK.setAccessToken(token)
    }
  }
}

File: flutter_app/android/app/src/main/kotlin/com/vuonrau/vuonrau/ezviz/EzvizSdk.kt
package com.vuonrau.vuonrau.ezviz

import android.app.Application
import com.videogo.openapi.EZOpenSDK
import java.net.URI

object EzvizSdk {
  @Volatile private var initialized = false
  private val lock = Any()

  private fun normalizeBaseUrl(raw: String?): String? {
    if (raw.isNullOrBlank()) return null
    return try {
      val uri = URI(raw)
      val scheme = uri.scheme ?: return raw
      val host = uri.host ?: return raw
      "$scheme://$host"
    } catch (_: Throwable) {
      raw
    }
  }

  fun initIfNeeded(
    application: Application,
    appKey: String,
    accessToken: String?,
    apiUrl: String?,
    authUrl: String?,
  ) {
    if (!initialized) {
      synchronized(lock) {
        if (!initialized) {
          // Must be before initLib.
          EZOpenSDK.showSDKLog(false)
          EZOpenSDK.enableP2P(true)
          EZOpenSDK.initLib(application, appKey)
          initialized = true
        }
      }
    }

    if (!accessToken.isNullOrBlank()) {
      EZOpenSDK.getInstance().setAccessToken(accessToken)
    }

    val normApiUrl = normalizeBaseUrl(apiUrl)
    val normAuthUrl = normalizeBaseUrl(authUrl)
    if (!normApiUrl.isNullOrBlank() && !normAuthUrl.isNullOrBlank()) {
      EZOpenSDK.getInstance().setServerUrl(normApiUrl, normAuthUrl)
    }
  }
}

